src/Core/Services/Implementations/OrganizationService.cs

```diff
@@ -1155,40 +1155,49 @@ public async Task<OrganizationUser> AcceptUserAsync(string orgIdentifier, User u
                 }
             }
 
-            ICollection<Policy> orgPolicies = null;
-            ICollection<Policy> userPolicies = null;
-            async Task<bool> hasPolicyAsync(PolicyType policyType, bool useUserPolicies = false)
-            {
-                var policies = useUserPolicies ? 
-                    userPolicies = userPolicies ?? await _policyRepository.GetManyByUserIdAsync(user.Id) : 
-                    orgPolicies = orgPolicies ?? await _policyRepository.GetManyByOrganizationIdAsync(orgUser.OrganizationId);
-                
-                return policies.Any(p => p.Type == policyType && p.Enabled);
-            }
-            var userOrgs = await _organizationUserRepository.GetManyByUserAsync(user.Id);
-            if (userOrgs.Any(ou => ou.OrganizationId != orgUser.OrganizationId && ou.Status != OrganizationUserStatusType.Invited))
-            {   
-                if (await hasPolicyAsync(PolicyType.SingleOrg))
-                {
-                    throw new BadRequestException("You may not join this organization until you leave or remove " +
-                        "all other organizations.");
-                }
-                if (await hasPolicyAsync(PolicyType.SingleOrg, true))
-                {
-                    throw new BadRequestException("You cannot join this organization because you are a member of " + 
-                        "an organization which forbids it");
-                }
+            bool notExempt(OrganizationUser organizationUser)
+            {
+                return organizationUser.Type != OrganizationUserType.Owner && organizationUser.Type != OrganizationUserType.Admin;
+            }
+
+            var allOrgUsers = await _organizationUserRepository.GetManyByUserAsync(user.Id);
+            var otherOrgUsers = allOrgUsers.Where(ou => ou.OrganizationId != orgUser.OrganizationId && ou.Status != OrganizationUserStatusType.Invited);
+
+            // Enforce Single Organization Policy of organization user is trying to join
+            var thisSingleOrgPolicy = await _policyRepository.GetByOrganizationIdTypeAsync(orgUser.OrganizationId, PolicyType.SingleOrg);
+            if (thisSingleOrgPolicy != null &&
+                thisSingleOrgPolicy.Enabled &&
+                notExempt(orgUser) &&
+                otherOrgUsers.Count() > 0)
+            {
+                throw new BadRequestException("You may not join this organization until you leave or remove " +
+                    "all other organizations.");
             }
 
-            if (!await userService.TwoFactorIsEnabledAsync(user))
+            // Enforce Single Organization Policy of other organizations user is a member of
+            foreach (var ou in otherOrgUsers)
```

Comment:
you can move the `notExtempt()` check inside this foreach block in a `.Where(o =>...` LINQ query as well to avoid additional nesting